### PR TITLE
Add new assist to convert for loop with while let 

### PR DESCRIPTION
**Pull request by arshiamoeini**
# Application

With this assist the users can convert their for to while let so that they can iterate their list manually.

For example, we have a text that we iterate by its iterator. If we come across the word a, then we are waiting for b after that, so we skip the iterator:

```rust
fn main() {
    let v = [1, 2, 3, 4];
    let mut v = v.into_iter();
    while let Some(x) = v.next() {
        if x == "a" {
            v.skip(1);
        }
    }
}
```

### CODE REVIEW

**Comment by ice1000** [comment]
Cool!!

File: crates/ide-assists/src/handlers/convert_iter_for_each_to_for.rs [hunk 1]
```diff
@@ -218,6 +221,19 @@ fn validate_method_call_expr(
     let iter_trait = FamousDefs(sema, krate).core_iter_Iterator()?;
     it_type.impls_trait(sema.db, iter_trait, &[]).then(|| (expr, receiver))
 }
+pub(crate) fn detect_for_loop(
```
Comment (jonas-schievink) [inline crates/ide-assists/src/handlers/convert_iter_for_each_to_for.rs:224]: This should probably be moved to some utils module, I don't think we should cross-import stuff between assist modules like that

**Review by jonas-schievink — COMMENTED** [review]
(did a first pass, don't have time for a full review right now, sorry)

File: crates/ide-assists/src/handlers/convert_iter_for_each_to_for.rs (hunk 1 above)
Comment (arshiamoeini) [reply crates/ide-assists/src/handlers/convert_iter_for_each_to_for.rs:224]: This function is from first few lines of "convert_for_loop_with_for_each". I'm not sure if I should copy it or put it into some function (here is detect_for_loop) and move it in utils.

**Comment by Veykril** [comment]
Can you remove the merge commits of the PR please? (Either by rebasing somehow or opening a new PR with the changes on a new branch) These merge commits create a lot of noise in the commit graph otherwise.

File: crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    acc.add(
+        AssistId("convert_for_loop_with_while_let", AssistKind::RefactorRewrite),
+        "Replace this for loop with while let",
+        for_loop.syntax().text_range(),
+        |builder| {
+            let mut buf = String::new();
+
+            let iterator_name;
+            println!("{}", iterable);
```
Comment (lnicola) [inline crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs:46]: Debugging leftovers?

File: crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    ast::{self},
+    AstNode,
+};
+
+use super::convert_iter_for_each_to_for::{impls_core_iter, is_ref_and_impls_iter_method};
+use crate::{utils::detect_for_loop, AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_for_loop_with_while_let
+//
+// Converts a for loop into a while let on the Iterator.
```
Comment (lnicola) [inline crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs:12]:
```suggestion
// Converts a `for` loop into a `while let` on the Iterator.
```

File: crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+                format_to!(
+                    buf,
+                    "let mut {} = {}.{}();\n    ",
+                    expr_behind_ref,
+                    expr_behind_ref,
+                    method
+                );
+                iterator_name = expr_behind_ref.to_string();
+            } else if let ast::Expr::RangeExpr(..) = iterable {
+                // range expressions need assigned with new variable
```
Comment (lnicola) [inline crates/ide-assists/src/handlers/convert_for_loop_with_while_let.rs:61]:
```suggestion
                // range expressions need to be bound to a new variable
```

**Comment by Veykril** [comment]
Closing this due to inactivity, feel free to re-open should you come back to it.

**Pull request closed without merging**